docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+# Incremental Generators Cookbook
+
+## Summary
+
+This document aims to be a guide to help the creation of source generators by providing a series of guidelines for common patterns.
+It also aims to set out what types of generators are possible under the current design, and what is expected to be explicitly out 
+of scope in the final design of the shipping feature.
+
+**This document expands on the details in the [full design document](incremental-generators.md), please ensure you have read that first.**
+
+## Table of content
+
+- [Incremental Generators Cookbook](#incremental-generators-cookbook)
+  - [Summary](#summary)
+  - [Table of content](#table-of-content)
+  - [Proposal](#proposal)
+  - [Out of scope designs](#out-of-scope-designs)
+    - [Language features](#language-features)
+    - [Code rewriting](#code-rewriting)
+  - [Conventions](#conventions)
+  - [Designs](#designs)
+    - [Generated class](#generated-class)
+    - [Additional file transformation](#additional-file-transformation)
+    - [Augment user code](#augment-user-code)
+    - [Issue Diagnostics](#issue-diagnostics)
+    - [INotifyPropertyChanged](#inotifypropertychanged)
+    - [Package a generator as a NuGet package](#package-a-generator-as-a-nuget-package)
+    - [Use functionality from NuGet packages](#use-functionality-from-nuget-packages)
+    - [Access Analyzer Config properties](#access-analyzer-config-properties)
+    - [Consume MSBuild properties and metadata](#consume-msbuild-properties-and-metadata)
+    - [Unit Testing of Generators](#unit-testing-of-generators)
+    - [Serialization](#serialization)
+    - [Auto interface implementation](#auto-interface-implementation)
+  - [Breaking Changes:](#breaking-changes)
+  - [Open Issues](#open-issues)
+
+## Proposal
+
+As a reminder, the high level design goals of source generators are:
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additive_ only. Generators can add new source code to a compilation but may **not** modify existing user code.
+- Can produce diagnostics. When unable to generate source, the generator can inform the user of the problem.
```

Comment:
note: this part is one i'm still oogy on since it often entirely breaks incremental.

Reply:
I did actually specifically say don't do diagnostics later on, I missed this one.